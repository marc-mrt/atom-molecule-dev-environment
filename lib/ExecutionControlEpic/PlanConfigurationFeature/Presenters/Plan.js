"use babel";
// @flow

import * as React from "react";
import type { TaskState } from "../../TaskExecutionFeature/Types/types.js.flow";
import DevToolCircle from "./DevToolCircle";
import Radium from "radium";
import { remote } from "electron";

const Menu = remote ? remote.Menu : undefined;
const MenuItem = remote ? remote.MenuItem : undefined;

let rippleAnimation = Radium.keyframes(
  {
    "0%": {
      transform: "scale(1)",
      background: "rgba(255,255,255,0)",
    },
    "50%": {
      background: "rgba(255,255,255,.2)",
    },
    "100%": {
      transform: "scale(30)",
      background: "rgba(255,255,255,0)",
    },
  },
  "ripple",
);

export function getStateColor(state: TaskState): string {
  switch (state) {
    case "running":
      return "#2EC4B6";
    case "failed":
      return "#E71D36";
    case "stopped":
      return "#1C1F26";
    case "succeed":
      return "#37CC44";
    case "crashed":
      return "#E71D36";
    case "created":
      return "#2EC4B6";
    default:
      return "#1C1F26";
  }
}

export class Plan extends React.Component<Props, State> {
  state: State;
  props: Props;
  static defaultProps: DefaultProps;

  constructor(props: Props) {
    super(props);
  }

  render() {
    return (
      <div
        className="plan-background-color"
        style={{
          margin: "0px 15px",
          fontSize: "20px",
          display: "inline-flex",
          padding: "15px 20px",
          transition: "border 1s",
          alignItems: "center",
          justifyContent: "center",
          borderRadius: "8px",
          cursor: "pointer",
          height: "50px",
          position: "relative",
          flexWrap: "no-wrap",
          flexGrow: "0",
          overflow: "hidden",
        }}
        onClick={() => this.props.onClick()}
        onContextMenu={() => {
          const menu = new Menu();
          menu.append(
            this.props.pinnable
              ? new MenuItem({
                  label: "Pin",
                  click: () => {
                    this.props.onPin();
                  },
                })
              : new MenuItem({
                  label: "Un-pin",
                  click: () => {
                    this.props.onUnpin();
                  },
                }),
          );
          menu.append(
            new MenuItem({
              label: "Remove",
              enabled: this.props.state !== "running",
              click: () => {
                this.props.onRemove();
              },
            }),
          );
          menu.popup(remote.getCurrentWindow());
        }}
      >
        <DevToolCircle
          iconUri={this.props.iconUri}
          color={getStateColor(this.props.state)}
        />
        <span
          style={{
            visibility: this.props.state == "running" ? "visible" : "hidden",
            animation: "x 2s ease 0s infinite",
            // Assign the result of `keyframes` to `animationName`
            animationName: rippleAnimation,
            borderRadius: "50%",
            height: "15px",
            width: "15px",
            position: "absolute",
            top: "20px",
            left: "35px",
          }}
        />
        <span
          className="text-color-highlight"
          style={{
            margin: "0px 15px",
            fontSize: "20px",
            fontFamily: "Inconsolata",
          }}
        >
          {this.props.name}
        </span>
      </div>
    );
  }
}

Plan.defaultProps = {
  pinnable: false,
  onClick: () => {},
};

type DefaultProps = {
  onClick(): void,
};

type Props = {
  pinnable: boolean,
  onPin(): void,
  onUnpin(): void,
  onRemove(): void,
  iconUri: string,
  name: string,
  state: TaskState,
  onClick(): void,
};

type State = {};

export default Radium(Plan);
