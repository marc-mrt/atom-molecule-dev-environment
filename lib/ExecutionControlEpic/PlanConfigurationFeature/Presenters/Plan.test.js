import "raf/polyfill";
import React from "react";
import Adapter from "enzyme-adapter-react-16";
import Enzyme, { mount } from "enzyme";
import Plan from "./Plan";
import { StyleRoot } from "radium";
import electron from "electron";

const remote = electron.remote;

Enzyme.configure({ adapter: new Adapter() });

beforeEach(() => {
  // $FlowFixMe
  remote.resetMenu();
});

describe("Plan", () => {
  it("should display the plan name", () => {
    let subject = mount(
      <StyleRoot>
        <Plan name="build" />
      </StyleRoot>,
    );

    expect(
      subject.findWhere(comp => comp.type() == "span" && comp.text() == "build")
        .length,
    ).toBe(1);
  });

  it("should display a context menu with Pin and Remove when pinnable is passed", () => {
    let subject = mount(
      <StyleRoot>
        <Plan pinnable />
      </StyleRoot>,
    );

    subject.find(Plan).simulate("contextmenu");
    expect(remote.getListReturn()).toMatchSnapshot();
  });

  it("should display a context menu with Unpin and Remove when pinnable is not passed", () => {
    let subject = mount(
      <StyleRoot>
        <Plan />
      </StyleRoot>,
    );

    subject.find(Plan).simulate("contextmenu");
    expect(remote.getListReturn()).toMatchSnapshot();
  });

  it("should display the tool icon", () => {
    let subject = mount(
      <StyleRoot>
        <Plan iconUri="atom://myplugin/icon.png" />
      </StyleRoot>,
    );

    expect(
      subject
        .find("img")
        .at(0)
        .prop("src"),
    ).toBe("atom://myplugin/icon.png");
  });

  it("should call onPin when Pin in context menu is clicked with pinnable", () => {
    let spy = jest.fn();
    let subject = mount(
      <StyleRoot>
        <Plan pinnable onPin={spy} />
      </StyleRoot>,
    );

    subject.find(Plan).simulate("contextmenu");
    remote.getListReturn()[0].click();
    expect(spy).toBeCalled();
  });

  it("should call onRemove when Remove in context menu is clicked with pinnable", () => {
    let spy = jest.fn();
    let subject = mount(
      <StyleRoot>
        <Plan pinnable onRemove={spy} />
      </StyleRoot>,
    );

    subject.find(Plan).simulate("contextmenu");
    remote.getListReturn()[1].click();
    expect(spy).toBeCalled();
  });

  it("should call onUnpin when Un-pin in context menu is clicked without pinnable", () => {
    let spy = jest.fn();
    let subject = mount(
      <StyleRoot>
        <Plan onUnpin={spy} />
      </StyleRoot>,
    );

    subject.find(Plan).simulate("contextmenu");
    remote.getListReturn()[0].click();
    expect(spy).toBeCalled();
  });

  it("should call onRemove when Remove in context menu is clicked without pinnable", () => {
    let spy = jest.fn();
    let subject = mount(
      <StyleRoot>
        <Plan onRemove={spy} />
      </StyleRoot>,
    );

    subject.find(Plan).simulate("contextmenu");
    remote.getListReturn()[1].click();
    expect(spy).toBeCalled();
  });

  it("should have Remove in context menu disabled while running", () => {
    let spy = jest.fn();
    let subject = mount(
      <StyleRoot>
        <Plan onRemove={spy} state="running" />
      </StyleRoot>,
    );

    subject.find(Plan).simulate("contextmenu");
    remote.getListReturn()[1].click();
    expect(remote.getListReturn()[1].enabled).toBeFalsy();
  });

  it("should have Remove in context menu enabled while not running", () => {
    let spy = jest.fn();
    let subject = mount(
      <StyleRoot>
        <Plan onRemove={spy} state="stopped" />
      </StyleRoot>,
    );

    subject.find(Plan).simulate("contextmenu");
    remote.getListReturn()[1].click();
    expect(remote.getListReturn()[1].enabled).toBeTruthy();
  });

  it("should call onClick when plan is clicked", () => {
    let spy = jest.fn();
    let subject = mount(
      <StyleRoot>
        <Plan onClick={spy} />
      </StyleRoot>,
    );

    subject.find(Plan).simulate("click");
    expect(spy).toBeCalled();
  });
});
