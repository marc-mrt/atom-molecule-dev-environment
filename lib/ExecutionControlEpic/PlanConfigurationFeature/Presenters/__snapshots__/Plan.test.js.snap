// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Plan should display a context menu with Pin and Remove when pinnable is passed 1`] = `
Array [
  MyMenuItem {
    "click": [Function],
    "enabled": undefined,
    "label": "Pin",
  },
  MyMenuItem {
    "click": [Function],
    "enabled": true,
    "label": "Remove",
  },
]
`;

exports[`Plan should display a context menu with Unpin and Remove when pinnable is not passed 1`] = `
Array [
  MyMenuItem {
    "click": [Function],
    "enabled": undefined,
    "label": "Un-pin",
  },
  MyMenuItem {
    "click": [Function],
    "enabled": true,
    "label": "Remove",
  },
]
`;
